{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bf66db8a-14cb-48cb-80af-db65213b402c",
   "metadata": {},
   "source": [
    "# A note on how to run AlphaFold 3\n",
    "\n",
    "**Author**: Xiping Gong (xipinggong@hotmail.com, Department of Food Science and Technology, College of Agricultural and Environmental Sciences, University of Georgia, Griffin, GA, USA)\n",
    "\n",
    "**Date**: 01/22/2025 (first draft); 04/24/2025 (create a modified \"af3.sh\" script and more)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a350859-bfb0-4294-89d3-9cd642ce1eb5",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "AlphaFold 2 has renolutionized biomolecular structrue prediction by providing accurate 3D protein structures, which can be effectively used for rapid molecular docking (DOI: https://doi.org/10.1038/s41586-021-03819-2). This year, AlphaFold 3 was launched, extending its capability to accurately model the biomolecule-ligand interactions, likely offering unprecedented precision in studying PFAS binding to critical toxicological targets, such as proteins (DOI: https://doi.org/10.1038/s41586-024-07487-w). It was claimed that its advanced predictive accuracy significantly surpasses that of tranditional molecular docking models (e.g., AutoDock Vina), providing more opportunities in understanding the PFAS-biomolecule binding mechanisms that drive PFAS bioaccumulation and toxicity (DOI: https://doi.org/10.1038/s41586-024-07487-w). The recent release of open-source code in November 2024 (Link: https://github.com/google-deepmind/alphafold3) introduces high-throughput capabilities, making it possible to rapidly screen a wide array of biomolecule-logand interactions. These advancements provide a foundation for generating high-quality structural features on PFAS-biomolecule interaction.\n",
    "\n",
    "This note uses the PFOA-human serum albumin interaction as an example to demonstrate how AlphaFold 3 can be utilized for docking. Additionally, I discuss the docking results and compare them to the outcomes obtained using AutoDock Vina from our previous note.\n",
    "\n",
    "AlphaFold 3: https://github.com/google-deepmind/alphafold3\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd4828e6-7b3c-4429-b72e-b7553f3bd027",
   "metadata": {},
   "source": [
    "# An example: PFOA - human serum albumin (hSA) protein\n",
    "\n",
    "The goal of this example is to how we can use the AlphaFold 3 to predict the binding of PFOA with the hSA protein. \n",
    "To test it, I integrated all scripts (Python and Bash) together, so that we can automatically screen other potential PFAS molecules.\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "dfaec45b-e5d8-428a-89b4-8087afe141a1",
   "metadata": {},
   "source": [
    "## Background\n",
    "\n",
    "**Reference**\n",
    "Maso, Lorenzo, et al. \"Unveiling the binding mode of perfluorooctanoic acid to human serum albumin.\" Protein Science 30.4 (2021): 830-841. DOI: https://doi.org/10.1002/pro.4036\n",
    "\n",
    "![Alt text](https://onlinelibrary.wiley.com/cms/asset/641b2e4e-b7a8-429b-8b78-d9238385a0ab/pro4036-fig-0001-m.jpg)\n",
    "\n",
    "**Figure 1**. Structure of hSA in complex with PFOA and Myr. Chemical structure (top) and composite omit maps depicting the (Foâˆ’Fc) electron density (bottom) of PFOA (a) and Myr (b) contoured at 4Ïƒ; (c) Crystal structure of hSA-PFOA-Myr complex (white) obtained using a twofold molar excess of PFOA over Myr [PDB identification code: 7AAI]; (d) Superimposition of hSA-PFOA-Myr ternary complex (white) with aligned hSA-Myr binary complex (blue white) [PDB identification code: 7AAE]. The structure of hSA is organized in homologues domains (I, II and III), subdomains (A and B), fatty acids (FA) and Sudlow's binding sites. The Î±-helices of hSA are represented by cylinders. Bound PFOA and Myr are shown in a ball-and-stick representation with a semi-transparent van der Waals and colored by atom type (PFOA: carbon = dark salmon, oxygen = firebrick, fluorine = palecyan; Myr: carbon = smudge green, oxygen = firebrick). The electron density PFOA and Myr is shown as grey mesh. (Note: I switched the \"7AAE\" with \"7AAI\" after checking out both structures from the PDB database.)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1aefe725-ab30-4aee-85fb-6981dd54b173",
   "metadata": {},
   "source": [
    "## Download the repository\n",
    "\n",
    "I have created a GitHub repository, and we should have it downloaded first, which includes some required scripts. \n",
    "\n",
    "```bash\n",
    "$ git clone https://github.com/XipingGong/pfas_docking.git\n",
    "$ cd pfas_docking # go to this directory, and we will have a test later.\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4178cf65-a81c-4c12-8403-f64e075e6573",
   "metadata": {},
   "source": [
    "## A general script to run the AF3 docking\n",
    "\n",
    "AF3 only requires a json file which includes the basic info of the protein-ligand complex, like protein sequence and ligand ID, so it can be straightforward to run the AlphaFold 3.\n",
    "\n",
    "+ **Step 1. Prepare the input files: af3.json and request the parameters file**\n",
    "\n",
    "A json example of hSA-PFOA is shown as follows, and save it as \"af3.json\". You can also check out the document for the details from here: https://github.com/google-deepmind/alphafold3/blob/main/docs/input.md\n",
    "\n",
    "```json\n",
    "{\n",
    "  \"name\": \"af3\",\n",
    "  \"sequences\": [\n",
    "    {\n",
    "    \"protein\": {\n",
    "        \"id\": \"A\",\n",
    "        \"sequence\": \"HKSEVAHRFKDLGEENFKALVLIAFAQYLQQCPFEDHVKLVNEVTEFAKTCVADESAENCDKSLHTLFGDKLCTVATLRETYGEMADCCAKQEPERNECFLQHKDDNPNLPRLVRPEVDVMCTAFHDNEETFLKKYLYEIARRHPYFYAPELLFFAKRYKAAFTECCQAADKAACLLPKLDELRDEGKASSAKQRLKCASLQKFGERAFKAWAVARLSQRFPKAEFAEVSKLVTDLTKVHTECCHGDLLECADDRADLAKYICENQDSISSKLKECCEKPLLEKSHCIAEVENDEMPADLPSLAADFVESKDVCKNYAEAKDVFLGMFLYEYARRHPDYSVVLLLRLAKTYETTLEKCCAAADPHECYAKVFDEFKPLVEEPQNLIKQNCELFEQLGEYKFQNALLVRYTKKVPQVSTPTLVEVSRNLGKVGSKCCKHPEAKRMPCAEDYLSVVLNQLCVLHEKTPVSDRVTKCCTESLVNRRPCFSALEVDETYVPKEFNAETFTFHADICTLSEKERQIKKQTALVELVKHKPKATKEQLKAVMDDFAAFVEKCCKADDKETCFAEEGKKLVAASQAAL\"\n",
    "      }\n",
    "    },\n",
    "    {\n",
    "     \"ligand\": {\n",
    "        \"id\": \"B\",\n",
    "        \"ccdCodes\": [\"8PF\"]\n",
    "      }\n",
    "    }\n",
    "  ],\n",
    "  \"modelSeeds\": [1],\n",
    "  \"bondedAtomPairs\": [],\n",
    "  \"dialect\": \"alphafold3\",\n",
    "  \"version\": 2\n",
    "}\n",
    "```\n",
    "\n",
    "+ **Step 2: Check out the \"af3.sh\" script**\n",
    "\n",
    "\n",
    "```bash\n",
    "# We provided a general script, \"af3.sh\", to run the AlphaFold 3.\n",
    "# Please modify the INPUT section in the \"scripts/af3.sh\" file, like\n",
    "# >> scripts_dir='/home/xg69107/program/pfas_docking/scripts' # need to change\n",
    "# >> af3_param_dir='/home/xg69107/program/alphafold3' # need to change\n",
    "# >> obabel=\"/home/xg69107/program/anaconda/anaconda3/envs/gmxMMPBSA/bin/obabel\"\n",
    "# >> python=\"/home/xg69107/program/anaconda/anaconda3/bin/python\"\n",
    "\n",
    "```\n",
    "\n",
    "+ **Step 3: Run the \"af3.sh\" script**\n",
    "\n",
    "```bash\n",
    "$ mkdir -p test/dock_dir/7AAI_8PF # create a test folder\n",
    "$ cd test/dock_dir/7AAI_8PF # go to this test folder\n",
    "$ sbatch ../../../scripts/af3.sh --input_json af3.json # submit an AF3 job and the results will be saved in an 'af3' folder.\n",
    "# It could take ~10 min if it runs\n",
    "\n",
    "```\n",
    "\n",
    "+ **Step 4: Check out the results**\n",
    "```bash\n",
    "$ ls -lrt af3 # see what this 'af3' folder has\n",
    "# >>\n",
    "# af3_data.json\n",
    "# af3_model.cif\n",
    "# af3_summary_confidences.json\n",
    "# af3_confidences.json\n",
    "# TERMS_OF_USE.md\n",
    "# ranking_scores.csv\n",
    "# best_pose\n",
    "# seed-1_sample-0\n",
    "# seed-1_sample-1\n",
    "# seed-1_sample-2\n",
    "# seed-1_sample-3\n",
    "# seed-1_sample-4\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efb389e0-4d42-4b7e-88b3-170307731db2",
   "metadata": {},
   "source": [
    "# Analysis & Conclusion\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00bf4b4e-239d-4e1d-a995-9f0aa7e1bd28",
   "metadata": {},
   "source": [
    "## RMSD check\n",
    "\n",
    "We can calcualte its RMSD value in terms of the native experimental structure.\n",
    "Usually, a RMSD value (<= 0.2 nm) can be a good prediction.\n",
    "\n",
    "+ **Obtain the native structure first**\n",
    "\n",
    "```bash\n",
    "# Download the native structure from the RCSB PDB\n",
    "$ bash ../../../scripts/get_native_pdb.sh --pdbid 7AAI --ligandid 8PF # it will generate four models because of four ligands\n",
    "# We can take the \"7AAI_8PF_1.pdb\" as the native structure\n",
    "$ cp 7AAI_8PF_1.pdb native_model.pdb\n",
    "```\n",
    "\n",
    "+ **Calculate the RMSD values of AF3-predicted structures**\n",
    "\n",
    "```bash\n",
    "# Before calculating the RMSD values, we still need to clean up the AF3-predicted structures, because they could have a different format from the native structure. We also need to do a structural alignment.\n",
    "$ bash ../../../scripts/af3_ana.sh # please have a look\n",
    "# >>\n",
    "# ðŸ“Š Protein Backbone RMSD (Direct): Min = 0.765 nm ; [0.79301757 0.7891323  0.78303146 0.7893516  0.79301757 0.7651273 ] nm\n",
    "# ðŸ“Š Protein Backbone RMSD (MDTraj): Min = 0.434 nm ; [0.43501106 0.44744506 0.4344369  0.45243776 0.43501106 0.4348642 ] nm\n",
    "# ðŸ“Š Ligand RMSD (Direct): Min = 0.846 nm ; [0.87349176 0.8715516  0.8641499  0.8538811  0.87349176 0.84595853] nm\n",
    "# ðŸ“Š Ligand RMSD (MDTraj): Min = 0.143 nm ; [0.14293166 0.17109454 0.1693879  0.17163862 0.14293166 0.164534  ] nm\n",
    "# ðŸ“Š Protein Backbone Pocket RMSD (Direct): Min = 0.077 nm ; [0.07853515 0.07989438 0.07657553 0.08200549 0.07853515 0.07801802] nm\n",
    "# ðŸ“Š Protein Backbone Pocket RMSD (MDTraj): Min = 0.077 nm ; [0.0785345  0.07989517 0.07657539 0.08200631 0.0785345  0.07801855] nm\n",
    "\n",
    "\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38435920-32cf-4c33-9d74-90db00ddfa85",
   "metadata": {},
   "source": [
    "## VMD visualization\n",
    "\n",
    "We can use the VMD software to have a look at the AF3-predicted structures, and see what protein residues interact with the PFOA molecule.\n",
    "\n",
    "Please see the demo (Link: ).\n",
    "\n",
    "```bash\n",
    "# Useful commands to obtain the hSA-PFOA interactions\n",
    "# Native structure\n",
    "$ python ../../../scripts/get_interactions_from_pdb.py native_model.pdb 8PF --cutoff_distance 0.3\n",
    "# AF3-predicted best pose\n",
    "$ python ../../../scripts/get_interactions_from_pdb.py af3/best_pose/aligned_model.pdb 8PF --cutoff_distance 0.3\n",
    "\n",
    "```\n",
    "\n",
    "<img src=\"af3_docking_pfoa_hsa.svg\" alt=\"Illustration of PFOA-hSA\" style=\"width:80%;\">\n",
    "\n",
    "**Figure 1** Comparison of PFOA-hSA interaction structures obtained experimentally and through AlphaFold 3 docking.\n",
    "\n",
    "The results reveal a close alignment between the two methods, with the head group of PFOA showing strong similarity. Notably, no specific binding pocket was predefined in this docking example, indicating that AlphaFold 3 can accurately predict the binding pocket of PFOA in the hSA protein. However, differences are observed in the orientation of the PFOA tail. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "475d7502-8885-4045-9d04-d453a0b6d4a5",
   "metadata": {},
   "source": [
    "# Questions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "370e19f3-2ba8-4bd7-adcb-a45f4b1a62a9",
   "metadata": {},
   "source": [
    "## How to submit an AF3 job at Sapelo2@GACRC?\n",
    "\n",
    "Please also check out the documentation from here: https://wiki.gacrc.uga.edu/wiki/AlphaFold3-Sapelo2 \n",
    "\n",
    "Please also check out the script 'af3.sh'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62dd80a9-6ac8-4b12-8df5-6e1108913ac9",
   "metadata": {},
   "source": [
    "## What if my ligand is negatively charged?\n",
    "\n",
    "The AF3 provides multiple ways to define the input ligand. For example, you can use a SMILES string to define the ligand structure, like the example below.\n",
    "Please see more details by checking this link: https://github.com/google-deepmind/alphafold3/blob/main/docs/input.md\n",
    "\n",
    "```bash\n",
    "{\n",
    "  \"name\": \"af3\",\n",
    "  \"sequences\": [\n",
    "    {\n",
    "      \"protein\": {\n",
    "        \"id\": \"A\",\n",
    "        \"sequence\": \"HKSEVAHRFKDLGEENFKALVLIAFAQYLQQCPFEDHVKLVNEVTEFAKTCVADESAENCDKSLHTLFGDKLCTVATLRETYGEMADCCAKQEPERNECFLQHKDDNPNLPRLVRPEVDVMCTAFHDNEETFLKKYLYEIARRHPYFYAPELLFFAKRYKAAFTECCQAADKAACLLPKLDELRDEGKASSAKQRLKCASLQKFGERAFKAWAVARLSQRFPKAEFAEVSKLVTDLTKVHTECCHGDLLECADDRADLAKYICENQDSISSKLKECCEKPLLEKSHCIAEVENDEMPADLPSLAADFVESKDVCKNYAEAKDVFLGMFLYEYARRHPDYSVVLLLRLAKTYETTLEKCCAAADPHECYAKVFDEFKPLVEEPQNLIKQNCELFEQLGEYKFQNALLVRYTKKVPQVSTPTLVEVSRNLGKVGSKCCKHPEAKRMPCAEDYLSVVLNQLCVLHEKTPVSDRVTKCCTESLVNRRPCFSALEVDETYVPKEFNAETFTFHADICTLSEKERQIKKQTALVELVKHKPKATKEQLKAVMDDFAAFVEKCCKADDKETCFAEEGKKLVAASQAAL\"\n",
    "      }\n",
    "    },\n",
    "    {\n",
    "      \"ligand\": {\n",
    "        \"id\": \"B\",\n",
    "        \"smiles\": \"O=C([O-])C(F)(F)C(F)(F)C(F)(F)C(F)(F)C(F)(F)C(F)(F)C(F)(F)F\"\n",
    "      }\n",
    "    }\n",
    "  ],\n",
    "  \"modelSeeds\": [\n",
    "    1\n",
    "  ],\n",
    "  \"bondedAtomPairs\": [],\n",
    "  \"dialect\": \"alphafold3\",\n",
    "  \"version\": 2\n",
    "}\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f52719b-4bf2-438b-b259-7774d2b0ab19",
   "metadata": {},
   "source": [
    "## If I have a new structure, then how can I align the AF3-predicted stuctures based on this new structure?\n",
    "\n",
    "```bash\n",
    "# We provide a simple bash script 'af3_ana.sh' to align the AF3-predicted structures based on any structure we have.\n",
    "# So, what you can do is just to replace the \"native_model.pdb\" file with the new structure file.\n",
    "#\n",
    "# Saved into an 'af3_ana.sh' bash file\n",
    "#\n",
    "scripts_dir=\"/home/xg69107/program/pfas_docking/scripts\"\n",
    "python=\"/home/xg69107/program/anaconda/anaconda3/bin/python\"\n",
    "\n",
    "NATIVE_MODEL=\"native_model.pdb\"\n",
    "NATIVE_LIGAND=\"native_ligand.pdb\"\n",
    "awk 'substr($0,22,1)==\"B\"' $NATIVE_MODEL > $NATIVE_LIGAND\n",
    "\n",
    "for bp in af3/*; do\n",
    "  [ -d \"$bp\" ] || continue\n",
    "\n",
    "  # split out chain A (protein) and chain B (ligand)\n",
    "  awk 'substr($0,22,1)==\"A\"' \"$bp/model.pdb\" > \"$bp/x_prot.pdb\"\n",
    "  awk 'substr($0,22,1)==\"B\"' \"$bp/model.pdb\" > \"$bp/x_lig.pdb\"\n",
    "\n",
    "  echo \"python $scripts_dir/rename_atom_info_by_topology.py --ref $NATIVE_LIGAND $bp/x_lig.pdb > $bp/x_lig_convert.pdb\"\n",
    "        python $scripts_dir/rename_atom_info_by_topology.py --ref $NATIVE_LIGAND $bp/x_lig.pdb > $bp/x_lig_convert.pdb\n",
    "  echo \"\"\n",
    "\n",
    "  # stitch protein + converted ligand into one PDB\n",
    "  cat \"$bp/x_prot.pdb\" \"$bp/x_lig_convert.pdb\" > \"$bp/model_convert.pdb\"\n",
    "\n",
    "  echo \"$bp: Converted $bp/model_convert.pdb\"\n",
    "done\n",
    "echo \"\"\n",
    "\n",
    "# Alignment: default 'aligned_model.pdb' will be created for each AF3-predicted structure\n",
    "$python $scripts_dir/align_pdb.py  --ref $NATIVE_MODEL \"af3/*/model_convert.pdb\"\n",
    "\n",
    "# Alignment: Convert all AF3-predicted structures into a pdb file 'af3_model.pdb'\n",
    "$python $scripts_dir/align_pdb.py  --ref $NATIVE_MODEL \"af3/*/aligned_model.pdb\" -o af3_model.pdb\n",
    "\n",
    "# Calculate RMSD in terms of native structure\n",
    "$python $scripts_dir/check_rmsd.py --ref $NATIVE_MODEL \"af3/*/aligned_model.pdb\"\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e73e211c-3842-418f-8ba8-b672086c4f9e",
   "metadata": {},
   "source": [
    "## How can save the RMSD data from multiple tests in a data file?\n",
    "\n",
    "```bash\n",
    "# Save the RMSD data into a data file\n",
    "$ python ../../../scripts/check_rmsd.py --ref native_model.pdb \"af3/*/aligned_model.pdb\" > af3_model.rmsd\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "```bash\n",
    "# If we have finished multiple tests and obtained many \"af3_model.rmsd\" data files, then we want to put these data together\n",
    "$ cd ../ # go back to a previous directory that includes many test examples\n",
    "$ python ../../scripts/extract_rmsd.py --num_per_array 6 \"*/af3_model.rmsd\"\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
